src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/LengthAttribute.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+
+namespace System.ComponentModel.DataAnnotations
+{
+    /// <summary>
+    ///     Specifies the minimum and maximum length of collection/string data allowed in a property.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Parameter, AllowMultiple = false)]
+    public class LengthAttribute : ValidationAttribute
+    {
+        [RequiresUnreferencedCode(CountPropertyHelper.RequiresUnreferencedCodeMessage)]
+        public LengthAttribute(int minimumLength, int maximumLength)
+            : base(SR.LengthAttribute_ValidationError)
+        {
+            MinimumLength = minimumLength;
+            MaximumLength = maximumLength;
+        }
+
+        /// <summary>
+        ///     Gets the minimum allowable length of the collection/string data.
+        /// </summary>
+        public int MinimumLength { get; }
+
+        /// <summary>
+        ///     Gets the maximum allowable length of the collection/string data.
+        /// </summary>
+        public int MaximumLength { get; }
+
+        /// <summary>
+        ///     Determines whether a specified object is valid. (Overrides <see cref="ValidationAttribute.IsValid(object)" />)
+        /// </summary>
+        /// <remarks>
+        ///     This method returns <c>true</c> if the <paramref name="value" /> is null.
+        ///     It is assumed the <see cref="RequiredAttribute" /> is used if the value may not be null.
+        /// </remarks>
+        /// <param name="value">The object to validate.</param>
+        /// <returns>
+        ///     <c>true</c> if the value is null or its length is between the specified minimum length and maximum length, otherwise
+        ///     <c>false</c>
+        /// </returns>
+        /// <exception cref="InvalidOperationException">
+        ///     <see cref="MinimumLength"/> is less than zero or <see cref="MaximumLength"/> is less than <see cref="MinimumLength"/>.
+        /// </exception>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "The ctor is marked with RequiresUnreferencedCode.")]
+        public override bool IsValid(object? value)
+        {
+            // Check the lengths for legality
+            EnsureLegalLengths();
```

Comment:
why we don't do that in the constructor?

Reply:
Good question. This appears to be standard practice in DataAnnotations attributes, see for example `MinLengthAttribute`:

https://github.com/dotnet/runtime/blob/71034dd2fbbd2304fac5c24d3a9f764a3c65f781/src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/MinLengthAttribute.cs#L47-L59

I'm guessing it's because the original authors would rather avoid throwing exceptions at attribute construction time and force these errors at validation time.

Reply:
We throw in the constructors in other cases, I guess. It looks weird to me to throw during the validation and not during the construction. I'll leave it to you to decide the best here.

Reply:
None of the other validation attributes appear to be throwing on the constructor. I'd probably have the constructor throw as well if I was writing this from scratch, but it's pretty clear that delayed argument validation was intentional (presumably because some validator component depends on validation attributes being queryable even if they are invalid? Maybe @jeffhandley knows)

I'd be inclined to preserve that invariant in the new attributes for the moment.

Reply:
@eiriktsarpalis It was certainly a rule in .NET Framework that the attributes needed to be constructable without throwing. As I remember it, this was for design-time scenarios in Visual Studio. I don't know if it is relevant anymore.

Reply:
I concur with the comments above that we tried to avoid throwing in constructors so that metadata models and design-time scenarios could construct the attributes without error. I also don't know to what degree that tenet is still relevant, but I'd be anxious about deviating from it without knowing how it would/could affect other systems.

Example scenario that is no longer applicable: Waaaaayyy back in the day, one of the scenarios was the build-time codegen we did for WCF RIA Services where we'd read the live code from an ASP.NET project and generate source files into a linked Silverlight project.